### ISSUE TITLE
Magpie can't always detect the game window is in foreground

### ISSUE THREAD

**Issue by rampaa** [Blinue/Magpie#538]
### Magpie version 程序版本

0.10.1

### Windows version 系统版本

Windows 10 22H2

### Related screenshot (optional) 相关截图（可选）

https://user-images.githubusercontent.com/25622653/229816615-72980c1b-db03-410a-8a7f-5ae6d6bc039d.mp4

### Reproduction steps 复现步骤

1) Add a Magpie profile for a game
2) Enable "Auto scale when in foreground"
3) Scale the game with Magpie
4) Use Windows+D (or Alt+Tab and bring another window to the foreground) to disable the scaling of the game window
5) Bring the game window back to the foreground by clicking its icon on the taskbar (or by using Alt+Tab again)

The game window will not be scaled by Magpie (seemingly randomly). You might need to re-try the 4th and 5th steps _many times_ before reproducing the issue because it happens quite randomly. It randomly happened with nearly every game I've tried Magpie with. "Capture method" doesn't seem to matter either, I can reproduce the issue with all the capture methods. I also don't think Magpie is producing any logs when it fails to detect that the game is in the foreground.

### Log files 日志文件

[magpie.log](https://github.com/Blinue/Magpie/files/11149570/magpie.log)

**Comment by Blinue** [comment]
Does v0.10.0 have this issue?

**Comment by rampaa** [comment]
Yes.

**Comment by Blinue** [comment]
Is there still an issue if you don't minimize the window (i.e., only use Alt+Tab)?

**Comment by rampaa** [comment]
I think I was able to reproduce the issue with Alt+Tab as well. To make sure I can try it out again when I am home.

**Comment by rampaa** [comment]
Reproducing it with Alt+Tab was significantly harder.

https://user-images.githubusercontent.com/25622653/230449822-448e0353-0290-4688-b063-f19f5579127c.mp4

**Comment by rampaa** [comment]
I'm not sure if this information will be helpful, but here it is anyway:

I recently tried the dev build (specifically, [Magpie-dev-97bdc7c-x64](https://github.com/Blinue/Magpie/actions/runs/15022512619)) to see if it would resolve the aforementioned issue. While it doesn’t seem to fix it, it does show the message "Scaling of this window is not supported." when the issue occurs, and it generates a related log. Here's the log file, in case it's useful:

[magpie.log](https://github.com/user-attachments/files/20213569/magpie.log)

**Comment by rampaa** [comment]
I've compiled Magpie and attempted to pinpoint the issue. It seems that, in my case, the source of the `ScalingError::InvalidSourceWindow` error is the following line:

https://github.com/Blinue/Magpie/blob/243649ed24b2ced9e94a560bef766c1cbdef6c9a/src/Magpie/ScalingService.cpp#L445

It appears that `GetWindowPlacement` sometimes incorrectly reports the window as minimized, even when it's not. As a result, Magpie concludes that the window isn’t suitable for scaling. Perhaps rechecking the window state with GetWindowPlacement to see if it still reports the window as minimized could resolve the issue?

**Comment by Blinue** [comment]
> It appears that GetWindowPlacement sometimes incorrectly reports the window as minimized, even when it's not. As a result, Magpie concludes that the window isn’t suitable for scaling. Perhaps rechecking the window state with GetWindowPlacement to see if it still reports the window as minimized could resolve the issue?

Makes sense, there could be an intermediate state during window restoration. Although the window becomes the foreground window, it may still be reported as minimized. The sequence might look like this:

User clicks taskbar icon → window becomes the "foreground window", but is still minimized → Magpie checks the foreground window and deems it unscalable → window finishes restoring

However, switching windows using Alt+Tab doesn’t involve a restoration process. Could you look into what’s happening in that case?

**Comment by rampaa** [comment]
> However, switching windows using Alt+Tab doesn’t involve a restoration process. Could you look into what’s happening in that case?

I can't reproduce the issue in the dev build when using Alt+Tab, since Magpie doesn't stop scaling the source window when another window becomes the foreground window. I tried stopping the scaling by pressing the "Stop Scaling" button and by using the scaling hotkey before switching with Alt+Tab, but neither method allowed me to reproduce the issue, possibly because I couldn't perform the actions rapidly enough.

In any case, I usually use Windows+D and rarely use Alt+Tab, so even if the issue still occurs with Alt+Tab, it's not a major inconvenience for me. If you could add more specific logs for the different reasons Magpie might stop scaling a window, that might help identify the underlying cause if anyone encounters the same issue again.

**Comment by Blinue** [comment]
Try [this version](https://github.com/Blinue/Magpie/actions/runs/15061910112), I believe the issue has been resolved.

> I can't reproduce the issue in the dev build when using Alt+Tab, since Magpie doesn't stop scaling the source window when another window becomes the foreground window.

Enabling 3D Game Mode will restore the old behavior, but I still haven’t been able to reproduce the issue. It’s possible that the Alt+Tab problem was inadvertently fixed, or that it only occurs under very specific and rare conditions.

**Comment by rampaa** [comment]
> Try [this version](https://github.com/Blinue/Magpie/actions/runs/15061910112), I believe the issue has been resolved.

I can no longer produce the issue, thanks!

> It’s possible that the Alt+Tab problem was inadvertently fixed, or that it only occurs under very specific and rare conditions.

Whichever the case may be, I cannot reproduce it any longer either.

**Comment by Blinue** [comment]
Fixed by https://github.com/Blinue/Magpie/commit/b73f19c52373a3ecd836fa798b929b8454aaca62.

> If I should open a separate issue for this, please let me know:
>
> There seems to be a regression in the dev builds compared to v0.11.2 that causes scaling to fail for some games when a game is first started.

Please file a new issue for this.
